docs/source/upgrade.rst

```diff
@@ -0,0 +1,747 @@
+.. currentmodule:: chainer
+
+Upgrade Guide from v1 to v2
+===========================
+
+This document provides detailed information for fixing your codes originally written for Chainer v1 to fit much better to Chainer v2.
+It also explains which part of your codes needs fixes when you upgrade Chainer from v1 to v2.
+
+.. contents::
+   :local:
+
+CuPy
+----
+
+.. _upgrade-cupy-separation:
+
+CuPy has been separated from Chainer into a separate package
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+CuPy, which was originally a part of Chainer, has been separated into a different Python package since Chainer v2.
+It changes the way to set up Chainer with CUDA support.
+In particular, you have to separately install :mod:`cupy` package to enable CUDA support.
+See :ref:`install-guide` for the recommended installation steps.
+
+Fortunately, there is no need of updating your source code to catch up with this change.
+
+
+Global configurations
+---------------------
+
+.. _upgrade-train-mode:
+
+Training mode is configured by a thread-local flag
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In Chainer v2, the concept of *training mode* is added.
+It is represented by a thread-local flag ``chainer.config.train``, which is a part of :ref:`the unified configuration <configuration>`.
+When ``chainer.config.train`` is true, functions of Chainer run in the training mode, and otherwise they run in the test mode.
+For example, :class:`~links.BatchNormalization` and :func:`~functions.dropout` behave differently in each mode.
+
+In Chainer v1, such a behavior was configured by the ``train`` argument of each function.
```

Comment:
Some links (e.g. `L.InceptionBN`) feed the state as `test` instead of `train`. So, it is better to mention `test` as well.